### PR TITLE
Some fixes to selector serialization re: namespaces and universal selector

### LINKED ISSUE #17182: Simplify `an+b` when serializing
When we serialize an `an+b` value, we just print it as `an+b`.

We should be simplifying it as much as possible. This means:

 - Printing out only `b` if `a` is zero
 - Printing out only `an` if `b` is zero
 - Printing `0` if both are zero
 - Not printing `a` if it is 1

Reduced testcase:

```html
<style type="text/css">
  :nth-child(+1n+5){}
</style>
<script type="text/javascript">document.write(document.styleSheets[0].cssRules[0].selectorText</script>
```

(Same for `+1n+0`, `+0n+5`, `+0n-0`, etc)

Relevant code: https://github.com/servo/servo/blob/2da8eb0342f6701b5365116b96113c83a1599b1f/components/selectors/parser.rs#L743

Testing: `tests/wpt/web-platform-tests/cssom/selectorSerialize.html`

### PR DESCRIPTION
**Pull request by jyc**
- Fix eliding default namespace when serializing
- Fix shortest serialization property when namespace prefix is `*|` and there is no default namespace
- Omit universal selector when serializing to match `cssom/serialize-namespaced-type-selectors` (again so we get the shortest serialization)

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes is part of a series to fix #17182

I'd like to land #17501 first, because it allows some tests for this to work.

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17537)

### CODE REVIEW

**Comment by emilio** [comment]
@bors-servo try

@jyc this looks good to me, is this waiting for anything in particular? Does it have tests?

**Comment by jyc** [comment]
Hi @emilio! Thanks, sorry for not being clear; was planning to run it to see what tests changed. It looks like it failed due to something in the infrastructure--

@bors-servo retry

**Comment by jdm** [comment]
```
     Running /home/travis/build/servo/servo/target/debug/deps/selectors-a25afd58e67e341d

running 11 tests

test parser::tests::test_empty ... ok

test parser::tests::test_empty_pseudo_iter ... ok

test parser::tests::test_pseudo_iter ... ok

test bloom::create_and_insert_some_stuff ... ok

test size_of_tests::size_of_component ... ok

test size_of_tests::size_of_pseudo_class ... ok

test size_of_tests::size_of_pseudo_element ... ok

test size_of_tests::size_of_selector ... ok

test parser::tests::visitor ... ok

test parser::tests::test_parsing ... FAILED

test parser::tests::test_universal ... FAILED
```

**Comment by jyc** [comment]
The tests used a parser with no default namespace but asserted that *| should be preserved in the output, which is contrary to the tests in `serialize-namespaced-type-selectors.htm`. Fixed those (and added tests for when there is a default namespace), and fixing the merge conflict now.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Universal selector in any namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by id

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by pseudo class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by pseudo element

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with any namespace followed by attribute selector

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Universal selector with namespace equal to default namespace

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by id

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by pseudo class

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by pseudo element

  ▶ Unexpected subtest result in /cssom/serialize-namespaced-type-selectors.html:
  └ PASS [expected FAIL] Type selector with namespace equal to default namespace followed by attribute selector
```

**Comment by jyc** [comment]
Thanks again! Also, I think it should be pretty straightforward to fix the remaining failures in that file. Working on it!

**Comment by jyc** [comment]
Created a spec issue at https://github.com/w3c/csswg-drafts/issues/1606 regarding what looks like an error in the spec vs. the web-platform test (which browsers pass).

**Comment by emilio** [comment]
@jyc when there's a new commit you need to use `try` instead of `retry` :P

**Comment by emilio** [comment]
```
	thread 'parser::tests::test_parsing' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/servo/buildbot/slave/linux-dev/build/components/selectors/parser.rs:776:59
```

**Comment by jyc** [comment]
@emilio Thanks! Makes sense.

**Comment by jyc** [comment]
Problem was the fake combinators we have for pseudo element selectors. Should be good now!

@bors-servo r?emilio

File: components/selectors/parser.rs [hunk 1]
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
...
+        let mut combinators = self.iter_raw_match_order().rev().filter(|x| x.is_combinator()).peekable();
         let compound_selectors = self.iter_raw_match_order().as_slice().split(|x| x.is_combinator()).rev();
 
         let mut combinators_exhausted = false;
         for compound in compound_selectors {
             debug_assert!(!combinators_exhausted);
-            for item in compound.iter() {
-                item.to_css(dest)?;
+
+            // https://www.w3.org/TR/cssom-1/#serializing-selectors
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: We usually link to `drafts.csswg.org`.

File: components/selectors/parser.rs [hunk 2]
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
...
+            //    which is a universal selector, append the result of
+            //    serializing the universal selector to s.
+            //
+            // Check if `compound.len() > 0` first--this can happen if we have
+            // something like `... > ::before`, because we store `>` and `::`
+            // both as combinators internally.
+            //
+            // If we are in this case, we continue to the next iteration of the
+            // `for compound in compound_selectors` loop.
+            let namespace = if compound.len() > 0 {
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: `if !compound.is_empty()`.

File: components/selectors/parser.rs [hunk 3]
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
...
+                let (namespace, first_non_namespace) = match &compound[0] {
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: Why the reference? I think it'd be clearer to do:

```rust
let has_namespace = matches!(compound[0], Component::ExplicitAnyNamespace | ...);
```

And derive the rest from that.

File: components/selectors/parser.rs [hunk 4]
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
...
+            //    that is not a universal selector of which the namespace prefix
+            //    maps to a namespace that is not the default namespace
+            //    serialize the simple selector and append the result to s.
+            //
+            // See https://github.com/w3c/csswg-drafts/issues/1606, which is
+            // proposing to change this to match up with the behavior asserted
+            // in cssom/serialize-namespaced-type-selectors.html, which the
+            // following code tries to match.
+            for simple in compound.iter() {
+                if let &Component::ExplicitUniversalType = simple {
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: we usually do:

```rust
if let Component::ExplicitUniversalType = *simple {
```

or,

```rust
if matches!(*simple, Component::ExplicitUniversalType) {
```

File: components/selectors/parser.rs [hunk 5]
```diff
@@ -763,19 +763,99 @@ impl<Impl: SelectorImpl> ToCss for Selector<Impl> {
...
+                    // Can't have a namespace followed by a pseudo-element
+                    // selector followed by a universal selector in the same
+                    // compound selector, so we don't have to worry about the
+                    // real namespace being in a different `compound`.
+                    match namespace {
+                        Some(&Component::DefaultNamespace(_)) | None => continue,
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: You can also use `matches!` here if you prefer, and avoid the `_ => ()` thing.

File: components/selectors/parser.rs [hunk 6]
```diff
@@ -1729,19 +1826,44 @@ pub mod tests {
...
+                    -> Result<SelectorList<DummySelectorImpl>, ParseError<'i, SelectorParseError<'i, ()>>>
+    {
+        parse_ns_expected(input, parser, None)
+    }
+
+    fn parse_ns_expected<'i, 'a>(
+        input: &'i str,
+        parser: &DummyParser,
+        expected: Option<&'a str>
+    ) -> Result<SelectorList<DummySelectorImpl>, ParseError<'i, SelectorParseError<'i, ()>>>
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: I think the brace should be in the previous line per https://github.com/rust-lang-nursery/fmt-rfcs

File: components/selectors/parser.rs [hunk 7]
```diff
@@ -2060,7 +2211,10 @@ pub mod tests {
 
     #[test]
     fn test_universal() {
-        let selector = &parse("*|*::before").unwrap().0[0];
+        let selector = &parse_ns("*|*::before", &DummyParser {
+            default_ns: Some(DummyAtom::from("https://mozilla.org")),
```
Comment (emilio) [inline components/selectors/parser.rs]: nit: Maybe having a `DummyParser::default()` or something like that would be nice, to avoid some duplication here.

**Review by emilio — COMMENTED** [review]
Looks good! Mostly nits.

Also maybe @SimonSapin wants to take a look.

File: components/selectors/parser.rs [hunk 8]
```diff
@@ -1732,19 +1746,44 @@ pub mod tests {
         }
     }
 
-    fn parse<'i>(input: &'i str) -> Result<SelectorList<DummySelectorImpl>,
-                                           ParseError<'i, SelectorParseError<'i, ()>>> {
+    fn parse<'i>(input: &'i str)
+                 -> Result<SelectorList<DummySelectorImpl>, ParseError<'i, SelectorParseError<'i, ()>>>
```
Comment (SimonSapin) [inline components/selectors/parser.rs]: No need to change this now, but next time you’d like to make whitespace-only changes please make them in a separate commit.

**Review by SimonSapin — COMMENTED** [review]
Looks fine to me, with @emilio’s comment addressed.

File: components/selectors/parser.rs (hunk 3 above)
Comment (jyc) [inline components/selectors/parser.rs]: Hm, I think it is clearer to keep the logic as close to where it is used as possible. Also, we don't quite want `has_namespace`, I think we instead want something like `can_elide_namespace`. Going to include proposed change, though.

File: components/selectors/parser.rs (hunk 4 above)
Comment (jyc) [inline components/selectors/parser.rs]: Didn't know about the `matches!` macro, thanks!

File: components/selectors/parser.rs (hunk 7 above)
Comment (jyc) [inline components/selectors/parser.rs]: Good idea!

**Comment by emilio** [comment]
@bors-servo r+

Awesome, thanks!

**Pull request merged**